### ISSUE TITLE
A New 20b128f Net(In Progress)

### ISSUE THREAD

**Issue by Grotex** [leela-zero/leela-zero#1168]
I use net2net get a 20b 128f network(un-trained):
fe0ad618
[https://mega.nz/#!Mzwn0BwS!i_sWP0TxECrHHOzvNcZ6kvtA5VJE59JrpFfUw2VIBD8](url)
According to my self-test its win-rate is only 51.25%(edit: should be 51.52%,a typo,sorry).
If you are interested, you can download it and train.
I am preparing trainingdata......
Grotex.

**Comment by hydrogenpi** [comment]
@Grotex thanks but the link is giving me a error at the moment? 
BTW while you are at it can you post a 40b 256f as well? Many thanks

**Comment by Grotex** [comment]
> what network did you net2net this from? the most recent 116 10block I assume?

yes
> Did you use the thing to make sure the heatmap and everything is aligned and working and not messed up? 

the heatmap is OK

>  if it isn't too much trouble if you can also provide and produce alongside the 20b 128f that you've already created (untrained) a 40b 256f (untrained) as well for those whom might want to go the largest size route that would be great. ;

if I have the rest time I would bring a 40b 256f(untrained)
> Based on your comment that you are "preparing trainingdata" my guess is that you might intend to go thorugh with actually training the net you justed created

yes. My data is from [https://sjeng.org/zero/](url)

**Comment by hydrogenpi** [comment]
Okay thanks, I'm assuming its the same as this link with the raw data : https://github.com/gcp/leela-zero/issues/167

I just about finished downloading all of it myself to preserve a local copies.

**Comment by isty2e** [comment]
Oh, I have been training a 128x20 net too. Can you provide training details, such as policy/MSE loss and accuracy?

**Comment by herazul** [comment]
Seems like GCP officially switched leela to 192*15, with some really good nets. It could be the new reference to test 20*256 or 20*128 net to see if there is improvement.

**Comment by Marcin1960** [comment]
"Seems like GCP officially switched leela to 192x15"


ba748d406 192x15  VS  39d46507 128x10 => 199 : 29 (87.28%) !!!!

Gian-Carlo,  "you have saved the best till now" :)

**Comment by isty2e** [comment]
Hmm, now it is in a awkward position... The motivation for 128x20 was speed for me, but the decision to switch to 192x15 being made, I wonder if 128x20 can be any useful.

**Comment by Marcin1960** [comment]
@isty2e 

Good news. The 192x15 is quite fast!

**Comment by hydrogenpi** [comment]
@Grotex I know you started with a  20b 128f , but since officially it is up to 15b 192f, would it make sense for you to use either the latest 15b 192f or the last/final 10b 128f to net2net to a 20b 192f or even a 20b 256f as opposed to a 20b 128f?

**Comment by Grotex** [comment]
@hydrogenpi 
OK I will "goto 20b 256f"
But my GPU will be the victim :-P

**Comment by Grotex** [comment]
Hey guys
we can discuss the net suited for every one's GPU

**Comment by Grotex** [comment]
But 20b 128f net I will train anyway
Look at its winrate

**Comment by Grotex** [comment]
Hmm
I will bring a 20b 256f and a 40b 256f, un-trained in another issue

**Comment by hydrogenpi** [comment]
okay thanks would very much appreciate a 40b 256f with good heatmaps! 

I think the stats on steam show a good snapshot at a glance as to what is the breakdown of current users and gpu type. But considering that GPU is improving all the time, what is the best fit today isn't going to be the best a year from now. Additionally, there can be different use cases. For example, during a match or special game what hardware will usually

**Comment by Grotex** [comment]
@hydrogenpi 
Thx for your hearten
My first step is to train 128x20 'cause my 256x40 will be come from that

**Comment by hydrogenpi** [comment]
Will the bulk of the games you use to train the 128*20 come from the 128*10 data? because if not, and you plan to use 192*15 to train the 128*20 wouldn't that be less effective? So I'm assuming the plan is to use the 128*10 games, several million historical ones, to train your 128*20 and then net2net the 128*20 into a 256*40 and then continue with the 40b training with the then official 192*15 games?

**Comment by Grotex** [comment]
@hydrogenpi 
Sorry for my omission
My data is from https://sjeng.org/zero/all_match.sgf.xz

**Comment by gcp** [comment]
>My data is from https://sjeng.org/zero/all_match.sgf.xz

Note that this is somewhat suboptimal as during the first 30 moves the move played may not necessarily be the best one (though it's proportionally so, so training might still work well enough). Raw data can be gotten from issue #167.

**Comment by isty2e** [comment]
@gcp A question here. If the trained 128x20 net is weaker than 192x15, which is highly probable, then it will not be promoted, right? Then I'm probably going to quit training.

**Comment by Grotex** [comment]
myplan:
128x10->128x20->256x20->256x40

**Comment by hydrogenpi** [comment]
I'm confused if this is not reinforcement learning and will only be supervised learning, why not just jump straight to the largest 256 * 40 and just keep up with the new games as they are generated?

**Comment by Grotex** [comment]
@hydrogenpi 
https://github.com/pangafu/LeelaMasterWeight/blob/master/How_to_Training_Yourself_Weight(CN).docx

**Comment by gcp** [comment]
> If the trained 128x20 net is weaker than 192x15, which is highly probable, then it will not be promoted, right?

No.

**Comment by Grotex** [comment]
@hydrogenpi 
No.I am training 20b 128f

**Comment by hydrogenpi** [comment]
Oh... Sorry to hear that. I use the LTSB of Win10 and turn of all updates in gpedit.msc

hope u can churn out a 40b soon. i would love to train it

**Comment by Grotex** [comment]
@hydrogenpi 
Thx.
when 20b 128f is done I will net2net to 40b 256f(but trainingwill be slower :-<)

**Comment by hydrogenpi** [comment]
ok nevermind, ill create my own 40b and jump straight into it then

**Comment by Grotex** [comment]
@hydrogenpi 
I find it out that my data has some weird move

>  Note that this is somewhat suboptimal as during the first 30
gcp is right

**Comment by Grotex** [comment]
I make fe0ad618 trained 128k step with lr=0.03, batch=1 and get this:
67982582
https://mega.nz/#!NjAVXCwB!rEnQy1hxV_iTMJbAvU-Jli8AIVAocRAngrsGju4aL4g
winrate is 50.19%
I wonder batch is major factor

**Comment by hydrogenpi** [comment]
you have to use the raw training data.... which is combined like almost 200 GB+

**Comment by Grotex** [comment]
@hydrogenpi 
using raw data will let my cp bomb(yes bomb,no burn) q_q

**Comment by Grotex** [comment]
@hydrogenpi 
I retrain fe0ad618 with lr=0.01, step=128k, batch=1
67982582 can take as a spare

**Comment by Grotex** [comment]
@hydrogenpi 
I will retrain fe0ad618 with lr=0.005, batch=1,step=128k
67982582 will be deleted

**Comment by hydrogenpi** [comment]
are you on windows or linux? please advise

**Comment by Grotex** [comment]
@hydrogenpi 
Windows

**Comment by hydrogenpi** [comment]
then contact me directly I can set you up on much better hardware for you to train,
as long as you agree to give me an untrained 40b 256 f first. 

bochen777@protonmail.com
972-737-7771

**Comment by Grotex** [comment]
@hydrogenpi 
Thx

**Comment by hydrogenpi** [comment]
text or email me and I can provide you credentials to remote into a windows server for training your stuff

**Comment by alreadydone** [comment]
@gcp https://github.com/gcp/leela-zero/issues/1168#issuecomment-379781459 Does this mean you no longer think training with only the most visited move is suboptimal (compared to the search probabilities)? @bjiyxo is apparently using raw data, but you think just the sgfs can achieve similar strength?

@hydrogenpi https://github.com/gcp/leela-zero/issues/1113#issuecomment-379742507 https://github.com/gcp/leela-zero/issues/1113#issuecomment-379804350 I think several net2net'd 192x15's should have been tested by now, but they are not much stronger than the randomly initialized ones. 
So starting from a net2net'd later network probably won't help much. 
256x40 definitely may be better. But doesn't net2net conversion just require one command line? And is it hard to launch `leelaz -w <weight file>` and type heatmap+Enter?

**Comment by hydrogenpi** [comment]
HOuston we have a problem...

![image](https://user-images.githubusercontent.com/35181396/38588886-9a5979e0-3d17-11e8-88f0-2089cc0d28ee.png)

**Comment by hydrogenpi** [comment]
New problem, the heat map is all wack!

![image](https://user-images.githubusercontent.com/35181396/38589445-84d7f986-3d1a-11e8-895c-fc4a4b642543.png)

**Comment by roy7** [comment]
You need to install the python module numpy. You could install a version of python for Windows that already has it included, or maybe this would help you. https://gehrcke.de/2015/02/how-to-set-up-a-64-bit-version-of-numpy-on-windows/

**Comment by hydrogenpi** [comment]
Training not working either....


![image](https://user-images.githubusercontent.com/35181396/38590779-dfa244d6-3d22-11e8-973a-bb02977595ff.png)

**Comment by Grotex** [comment]
@hydrogenpi 
Sorry to hear that
I have sent you a mail and a 40b 256f net
Be positive:-)

**Comment by bjiyxo** [comment]
@hydrogenpi Which version of tensorflow, cuda, cudnn are you using? Be aware of that tensorflow 1.4.0 doesn't support python 3.6.

**Comment by hydrogenpi** [comment]
@bjiyxo thanks, this is a tensorflow and python version mismatch... so I need to find the right version combinations and try again for the training part. 

Thanks 阿尔克杰诺 aka @Grotex I got your 40b 256f and can confirm its a good net, and its working fine! No heatmap issues. Now it just needs to be trained to be stronk! lol

**Comment by Grotex** [comment]
@hydrogenpi 
you are welcome
BTW I am Chinese

**Comment by hydrogenpi** [comment]
Ok hope you got my email and that the information worked for you!

**Comment by Grotex** [comment]
@hydrogenpi 
But how to install windows server

**Comment by hydrogenpi** [comment]
Its already installed, just rdp to it, I sent you email with ip address and username and password.

**Comment by hydrogenpi** [comment]
in windows machine or windows VM, go to start -> run -> mstsc

**Comment by hydrogenpi** [comment]
![image](https://user-images.githubusercontent.com/35181396/38593539-b087636a-3cf6-11e8-93b9-87725e21f3cb.png)

**Comment by Grotex** [comment]
@hydrogenpi 
Ok I get it
where is training dir

**Comment by hydrogenpi** [comment]
Its not there, you have to do it yourself. I coulndn't even figure out h ow to install the right versions of tensorflow / python etc...

**Comment by Grotex** [comment]
@hydrogenpi 
Thx for your rdp

**Comment by hydrogenpi** [comment]
@grotex no problem, thanks for the 40b 256f. step in the right direction. now all I have to do is training it to make it strong

**Comment by Grotex** [comment]
@hydrogenpi 
Your rdp is unstable q_q

**Comment by Grotex** [comment]
https://mega.nz/#!A2AA2BwS!uBa4uoh09ql6tNCKX2ER44jTNzu5UUlKminRlWX6_8Q
my data

**Comment by Grotex** [comment]
@hydrogenpi 
in your rdp I can't install tensorflow
Then I completely can't train

**Comment by Grotex** [comment]
@hydrogenpi 
Thx Again
I can train my fe0ad618  with step=128k, lr=0.03, batch=512

**Comment by Grotex** [comment]
@hydrogenpi 
Oh no I want to restart your rdp but

**Comment by alreadydone** [comment]
https://github.com/gcp/leela-zero/blob/872d4bb5fc5bbd6d02016c6a5e571796f44b985d/training/tf/net2net.py#L129

@Ttl As observed by @hydrogenpi above, this completely breaks the heatmap for a nonzero-bias net; this could partly explain why @gcp's net2net'd 192x15 (from 2e3863ed) doesn't show difference with random initialization.

An accidental discovery was that leelaz doesn't accept 194 or 196 filters; it spits out the error
`Error in convolve1: clEnqueueNDRangeKernel: -54`
Does the number of filters have to be a multiple of something?

@Grotex I think your 256x40 was net2net'd from 256x20, right? Which one was it?

**Comment by Grotex** [comment]
1629cbfd
@alreadydone

**Comment by gcp** [comment]
>Does the number of filters have to be a multiple of something?

Yes, 8. A limitation of the current kernel for 1x1 convolutions. Other sizes might work with the CPU only version.

**Comment by gcp** [comment]
>this could partly explain why @gcp's net2net'd 192x15 (from 2e3863ed) doesn't show difference with random initialization

Only partly though. IIRC the beta parameters that are zeroed here are very small in practice, so it's something training should very quickly fix.

**Comment by Grotex** [comment]
@hydrogenpi 
I install tensorflow but it don't work q_q

**Comment by hydrogenpi** [comment]
@Grotex rotex Okay but I just restarted it for you.

**Comment by Ttl** [comment]
@alreadydone @hydrogenpi Thanks for the bug report. Fixed in #1187.

> IIRC the beta parameters that are zeroed here are very small in practice, so it's something training should very quickly fix.

In the latest 192x15 networks most of them are actually very big. First convolution in the residual block seems to have almost all biases about -1 to -2, when they were about -0.005 in the 128x10 networks.

**Comment by gcp** [comment]
>In the latest 192x15 networks most of them are actually very big

Hmm, being biased variables they're not covered by regularization. If they're very negative, that means they're pulling more things into the negative zone (0) of the ReLU?

**Comment by Grotex** [comment]
@hydrogenpi 
Hello bochen
I am back
I successful run training of fe0ad618 in phase 1:step=128k, lr=0.03, batch=1
So I am goto phase2:step=128k, lr=0.01, battch=1

**Comment by Grotex** [comment]
phase3 done
phase4 step=128k, lr=0.001, batch=1

**Comment by Grotex** [comment]
phase4 done
phase5 step=128k, lr=0.0005, batch=1

**Comment by Grotex** [comment]
phase4 done
I will give the trained net later

**Comment by Grotex** [comment]
@hydrogenpi 
Hi I quit supervised train and go to rest
I will be absent in days

**Comment by Grotex** [comment]
@hydrogenpi 
I don't have 40 block to train except the one I gave you

**Comment by Grotex** [comment]
> Hi I quit supervised train and go to rest
I will be absent in days

I am back
but already absented in months
sorry
and then I get a 20b 256f (b1c37640) and net2net to 40b 256f
the next step is get data
@hydrogenpi 
how do you do

**Comment by l1t1** [comment]
what size of weight can get by net2net?
224x20?
256x19?

**Comment by Grotex** [comment]
> what size of weight can get by net2net?
224x20?
256x19?

net2net can make a small net bigger
e.g. 192x20 -> 224x20
but 256x20 -/> 224x20
224x40 -/>  224x20

**Comment by l1t1** [comment]
thanks, could you make a 256x20 using elfv1( 224x20)? @Grotex

**Comment by Grotex** [comment]
where is elfv1
@l1t1

**Comment by l1t1** [comment]
elfv1 http://zero.sjeng.org/networks/d13c40993740cb77d85c838b82c08cc9c3f0fbc7d8c3761366e5d59e8f371cbd.gz
and I want a converted 256x20 weight file to test.

**Comment by roy7** [comment]
Remember you can't train with ELF nets because their win rate data is opposite what LZ uses.

**Comment by alreadydone** [comment]
There are already strong 40b networks trained by bjiyxo available at zero.sjeng.org. He used Xavier initialization instead of net2net IIRC.

I don't remember how people concluded it's hard to train from ELF, but I think we just need revert the winner (1 for side-to-move and -1 for the opposite side in the training data) when white is the side to move, and one more script to modify the training data in the following manner should do the task:
+ If line 19n+17 is 1, then change the sign of line 19n+19.

The training code doesn't support sampling 50% black-won games and 50% white-won games for training though, which the ELF team has been doing, for inexplicable reason to me.

Also the current training data format is suitable for training with different komi; currently a 0 on line 19n+17 means komi is 7.5 for the side to move, and a 1 means it's -7.5 for the side to move, and we can easily replace it by any floating point number.

**Comment by sethtroisi** [comment]
Closing older issues.
We've had 40b and many 20b since this :)
